apps/ogr2ogr_lib.cpp

```diff
@@ -5663,46 +5680,133 @@ int LayerTranslator::Translate(OGRFeature *poFeatureIn, TargetLayerInfo *psInfo,
                                 }
                             }
+
+                            CPLError(CE_Failure, CPLE_AppDefined,
+                                     "Failed to reproject feature " CPL_FRMT_GIB
+                                     " (geometry probably out of source or "
+                                     "destination SRS).",
+                                     nSrcFID);
+                            if (!psOptions->bSkipFailures)
+                            {
+                                delete poDstGeometry;
+                                return false;
+                            }
                         }
 
-                        CPLError(CE_Failure, CPLE_AppDefined,
-                                 "Failed to reproject feature " CPL_FRMT_GIB
-                                 " (geometry probably out of source or "
-                                 "destination SRS).",
-                                 nSrcFID);
-                        if (!psOptions->bSkipFailures)
+                        // Check if a curve geometry is no longer valid after
+                        // reprojection
+                        const auto eType = poDstGeometry->getGeometryType();
+                        const auto eFlatType = wkbFlatten(eType);
+
+                        const auto IsValid = [](const OGRGeometry *poGeom)
+                        {
+                            CPLErrorHandlerPusher oErrorHandler(
+                                CPLQuietErrorHandler);
+                            return poGeom->IsValid();
+                        };
+
+                        if (iIter == 0 && bReprojCanInvalidateValidity &&
+                            OGRGeometryFactory::haveGEOS() &&
+                            (eFlatType == wkbCurvePolygon ||
+                             eFlatType == wkbCompoundCurve ||
+                             eFlatType == wkbMultiCurve ||
+                             eFlatType == wkbMultiSurface) &&
+                            poDstGeometry->hasCurveGeometry(TRUE) &&
+                            IsValid(poDstGeometry))
+                        {
+                            OGRwkbGeometryType eTargetType = OGR_GT_GetLinear(
+                                poDstGeometry->getGeometryType());
+                            auto poDstGeometryTmp =
+                                std::unique_ptr<OGRGeometry>(
+                                    OGRGeometryFactory::forceTo(
+                                        poReprojectedGeom->clone(),
+                                        eTargetType));
+                            if (!IsValid(poDstGeometryTmp.get()))
+                            {
+                                CPLDebug("OGR2OGR",
+                                         "Curve geometry no longer valid after "
+                                         "reprojection: transforming it into "
+                                         "linear one before reprojecting");
```

Comment:
For this [data](https://share.krzn.de/alfresco/service/ecm4u/exs/kreis-viersen/share/9URfvgdO/download/d29ya3NwYWNlOi8vU3BhY2VzU3RvcmUvYzkyZmUzMmMtYzI3YS00MjIzLTkxMDgtOGU5YTRjOTQyNzIx/test2.gml) with
`ogr2ogr -f GeoJSON out2_rfc.geojson test2.gml -lco RFC7946=YES --debug on`
there are 41 invalid geometries in the GeoJSON.

I suspect this is because of the default precision of 7 for RFC7946 because with 
`ogr2ogr -f GeoJSON out2_rfc_prec_15.geojson test2.gml -lco RFC7946=YES -lco COORDINATE_PRECISION=15`
all resulting geometries are valid.

But shouldn't be there this debug output?
I only get 
```
ogr2ogr -f GeoJSON out2_rfc.geojson test2.gml -lco RFC7946=YES --debug on
GML: Using Expat reader
GML: Global SRS = urn:ogc:def:crs:EPSG::25832
GDAL: GDALOpen(test2.gml, this=0x56267279e8e0) succeeds as GML.
GDAL: QuietDelete(out2_rfc.geojson) invoking Delete()
GDAL: GDALDriver::Create(GeoJSON,out2_rfc.geojson,0,0,0,Unknown,(nil))
GML: ResetReading()
GML: ResetReading()
GDALVectorTranslate: 176666 features written in layer 'test2'
GDAL: GDALClose(test2.gml, this=0x56267279e8e0)
GDAL: GDALClose(out2_rfc.geojson, this=0x562672b12120)
GDAL: In GDALDestroy - unloading GDAL shared library.
```

Reply:
> But shouldn't be there this debug output?

not one about geometry type conversion, since this is expected given the capabilities of the driver.
That said I've added a new commit to deal with the reduced coordinate precision of GeoJSON output to run MakeValid() if the input was valid, but the output (after rounding, and before this new correction) was invalid. Works well on your test dataset

Reply:
Nice!

Works well also for other big datasets I've tried :tada: